storage/src/merkle_storage_stats.rs

```diff
@@ -0,0 +1,256 @@
+    pub global: OperationLatencyStats,
+    pub perpath: PerPathOperationStats,
+}
+
+#[derive(Serialize, Default, Debug, Clone)]
+pub struct MerkleStorageStatistics {
+    pub perf_stats: MerklePerfStats,
+    pub block_latencies: BlockLatencies,
+}
+
+impl BlockLatencies {
+    fn update(&mut self, latency: u64) {
+        self.current += latency;
+    }
+
+    fn end_block(&mut self) {
+        self.latencies.push(self.current);
+        self.current = 0;
+    }
+
+    pub fn get(&self, offset_from_last_applied: usize) -> Option<u64> {
+        self.latencies
+            .len()
+            .checked_sub(offset_from_last_applied + 1)
+            .and_then(|index| self.latencies.get(index))
+            .copied()
+    }
+}
+
+pub enum MerkleStorageAction {
+    Set,
+    Get,
+    GetByPrefix,
+    GetKeyValuesByPrefix,
+    GetContextTreeByPrefix,
+    GetHistory,
+    Mem,
+    DirMem,
+    Copy,
+    Delete,
+    DeleteRecursively,
+    Commit,
+    Checkout,
+    GarbageCollector,
+}
+
+impl fmt::Display for MerkleStorageAction {
+    fn fmt(&self, f: &mut fmt::Formatter) -> fmt::Result {
+        match self {
+            MerkleStorageAction::Set => {
+                write!(f, "Set").unwrap();
+            }
+            MerkleStorageAction::Get => {
+                write!(f, "Get").unwrap();
+            }
+            MerkleStorageAction::GetByPrefix => {
+                write!(f, "GetByPrefix").unwrap();
+            }
+            MerkleStorageAction::GetKeyValuesByPrefix => {
+                write!(f, "GetKeyValuesByPrefix").unwrap();
+            }
+            MerkleStorageAction::GetContextTreeByPrefix => {
+                write!(f, "GetContextTreeByPrefix").unwrap();
+            }
+            MerkleStorageAction::GetHistory => {
+                write!(f, "GetHistory").unwrap();
+            }
+            MerkleStorageAction::Mem => {
+                write!(f, "Mem").unwrap();
+            }
+            MerkleStorageAction::DirMem => {
+                write!(f, "DirMem").unwrap();
+            }
+            MerkleStorageAction::Copy => {
+                write!(f, "Copy").unwrap();
+            }
+            MerkleStorageAction::Delete => {
+                write!(f, "Delete").unwrap();
+            }
+            MerkleStorageAction::DeleteRecursively => {
+                write!(f, "DeleteRecursively").unwrap();
+            }
+            MerkleStorageAction::Commit => {
+                write!(f, "Commit").unwrap();
+            }
+            MerkleStorageAction::Checkout => {
+                write!(f, "Checkout").unwrap();
+            }
+            MerkleStorageAction::GarbageCollector => {
+                write!(f, "GarbageCollector").unwrap();
+            }
+        };
+        Ok(())
+    }
+}
+
+pub struct StatUpdater<'a> {
+    stats: &'a mut MerkleStorageStatistics,
+    timer: Instant,
+    action: MerkleStorageAction,
+    key: Option<String>,
+}
+
+impl<'a> Drop for StatUpdater<'a> {
```

Comment:
Cool. Just in case, we want these stats even if the operation fails, right? (which is unlike the previous implementation)